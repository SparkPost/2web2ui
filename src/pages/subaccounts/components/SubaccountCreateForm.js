import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm, formValueSelector } from 'redux-form';
import { Button, Panel, Stack } from 'src/components/matchbox';
import { PageLink } from 'src/components/links';
import { Form } from 'src/components/form';
import { ButtonWrapper } from 'src/components';
import { getSubaccountGrants, getInitialSubaccountGrants } from 'src/selectors/api-keys';
import { getIpPools, selectFirstIpPoolId } from 'src/selectors/ipPools';
import { NameField, ApiKeyCheckBox, ApiKeyFields } from './formFields';
import IpPoolSelect from './IpPoolSelect';
import RestrictToIpPoolCheckbox from './RestrictToIpPoolCheckbox';
export class SubaccountCreateForm extends Component {
  render() {
    const {
      handleSubmit,
      pristine,
      showGrants,
      grants,
      submitting,
      createApiKey,
      ipPools,
      restrictedToIpPool,
    } = this.props;

    return (
      <Form onSubmit={handleSubmit} id="subaccount-create-form">
        <Panel.LEGACY.Section>
          <Stack>
            <NameField disabled={submitting} />
            <ApiKeyCheckBox disabled={submitting} createApiKey={createApiKey} />
            <ApiKeyFields
              show={createApiKey}
              showGrants={showGrants}
              grants={grants}
              submitting={submitting}
            />
            {Boolean(ipPools.length) && (
              <Stack>
                <RestrictToIpPoolCheckbox disabled={submitting} />
                {restrictedToIpPool && <IpPoolSelect disabled={submitting} options={ipPools} />}
              </Stack>
            )}
          </Stack>
        </Panel.LEGACY.Section>
        <Panel.LEGACY.Section>
          <ButtonWrapper>
            <Button variant="primary" submit disabled={submitting || pristine}>
              {submitting ? 'Loading...' : 'Create Subaccount'}
            </Button>

            <PageLink
              as={Button}
              variant="secondary"
              to="/account/subaccounts"
              disabled={submitting}
            >
              Cancel
            </PageLink>
          </ButtonWrapper>
        </Panel.LEGACY.Section>
      </Form>
    );
  }
}

const formName = 'SubaccountCreateForm';
const valueSelector = formValueSelector(formName);

const mapStateToProps = state => ({
  grants: getSubaccountGrants(state),
  createApiKey: valueSelector(state, 'createApiKey'),
  showGrants: valueSelector(state, 'grantsRadio') === 'select',
  ipPools: getIpPools(state),
  restrictedToIpPool: valueSelector(state, 'restrictedToIpPool'),
  initialValues: {
    grants: getInitialSubaccountGrants(state),
    grantsRadio: 'all',
    createApiKey: true,
    restrictedToIpPool: false,
    ipPool: selectFirstIpPoolId(state),
  },
});

const SubaccountReduxForm = reduxForm({ form: formName })(SubaccountCreateForm);
export default connect(mapStateToProps, {})(SubaccountReduxForm);
