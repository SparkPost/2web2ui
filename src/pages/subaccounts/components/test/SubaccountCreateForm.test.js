import { shallow } from 'enzyme';
import React from 'react';

import { SubaccountCreateForm } from '../SubaccountCreateForm';

const handleSubmit = jest.fn();

const props = {
  grants: ['subaccount grants'],
  showGrants: false,
  createApiKey: false,
  ipPools: [],
  handleSubmit
};

let wrapper;

beforeEach(() => {
  wrapper = shallow(<SubaccountCreateForm {...props} />);
});

test('base case', () => {
  expect(wrapper).toMatchSnapshot();
});

test('api key Form', () => {
  wrapper.setProps({ createApiKey: true });
  expect(wrapper).toMatchSnapshot();
});

test('has ip pools', () => {
  wrapper.setProps({ ipPools: ['salt water']});
  expect(wrapper).toMatchSnapshot();
});

test('has ip pools and checked checkbox', () => {
  wrapper.setProps({ ipPools: ['salt water'], restrictedToIpPool: true });
  expect(wrapper).toMatchSnapshot();
});

test('submits correctly', () => {
  wrapper.find('Form').simulate('submit');
  expect(handleSubmit).toHaveBeenCalled();
  wrapper.setProps({ submitting: true });
  expect(wrapper).toMatchSnapshot();
});
