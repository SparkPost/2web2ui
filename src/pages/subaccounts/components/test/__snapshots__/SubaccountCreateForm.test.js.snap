// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api key Form 1`] = `
<Form
  id="subaccount-create-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <NameField />
      <ApiKeyCheckBox
        createApiKey={true}
      />
      <ApiKeyFields
        grants={
          Array [
            "subaccount grants",
          ]
        }
        show={true}
        showGrants={false}
      />
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        submit={true}
        variant="primary"
      >
        Create Subaccount
      </Button>
      <PageLink
        as={[Function]}
        to="/account/subaccounts"
        variant="secondary"
      >
        Cancel
      </PageLink>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`base case 1`] = `
<Form
  id="subaccount-create-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <NameField />
      <ApiKeyCheckBox
        createApiKey={false}
      />
      <ApiKeyFields
        grants={
          Array [
            "subaccount grants",
          ]
        }
        show={false}
        showGrants={false}
      />
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        submit={true}
        variant="primary"
      >
        Create Subaccount
      </Button>
      <PageLink
        as={[Function]}
        to="/account/subaccounts"
        variant="secondary"
      >
        Cancel
      </PageLink>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`has ip pools 1`] = `
<Form
  id="subaccount-create-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <NameField />
      <ApiKeyCheckBox
        createApiKey={false}
      />
      <ApiKeyFields
        grants={
          Array [
            "subaccount grants",
          ]
        }
        show={false}
        showGrants={false}
      />
      <Stack>
        <RestrictToIpPoolCheckbox />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        submit={true}
        variant="primary"
      >
        Create Subaccount
      </Button>
      <PageLink
        as={[Function]}
        to="/account/subaccounts"
        variant="secondary"
      >
        Cancel
      </PageLink>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`has ip pools and checked checkbox 1`] = `
<Form
  id="subaccount-create-form"
  onSubmit={[MockFunction]}
>
  <Panel.LEGACY.Section>
    <Stack>
      <NameField />
      <ApiKeyCheckBox
        createApiKey={false}
      />
      <ApiKeyFields
        grants={
          Array [
            "subaccount grants",
          ]
        }
        show={false}
        showGrants={false}
      />
      <Stack>
        <RestrictToIpPoolCheckbox />
        <IpPoolSelect
          options={
            Array [
              "salt water",
            ]
          }
        />
      </Stack>
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        submit={true}
        variant="primary"
      >
        Create Subaccount
      </Button>
      <PageLink
        as={[Function]}
        to="/account/subaccounts"
        variant="secondary"
      >
        Cancel
      </PageLink>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;

exports[`submits correctly 1`] = `
<Form
  id="subaccount-create-form"
  onSubmit={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": undefined,
    }
  }
>
  <Panel.LEGACY.Section>
    <Stack>
      <NameField
        disabled={true}
      />
      <ApiKeyCheckBox
        createApiKey={false}
        disabled={true}
      />
      <ApiKeyFields
        grants={
          Array [
            "subaccount grants",
          ]
        }
        show={false}
        showGrants={false}
        submitting={true}
      />
    </Stack>
  </Panel.LEGACY.Section>
  <Panel.LEGACY.Section>
    <ButtonWrapper>
      <Button
        disabled={true}
        submit={true}
        variant="primary"
      >
        Loading...
      </Button>
      <PageLink
        as={[Function]}
        disabled={true}
        to="/account/subaccounts"
        variant="secondary"
      >
        Cancel
      </PageLink>
    </ButtonWrapper>
  </Panel.LEGACY.Section>
</Form>
`;
